File: python/helpers/stateEstimation.py
##############################################################################
# State estimation in LGSS and SV models using Kalman and particle filters
# (c) Johan Dahlin 2017 under MIT license <liu@johandahlin.com.nospam>
##############################################################################

from __future__ import print_function, division
import numpy as np
from numpy.random import randn, choice
from scipy.stats import norm

##############################################################################
# Kalman filter for the linear Gaussian SSM
##############################################################################
def kalmanFilter(observations, parameters, initialState, initialStateCov):
    
    noObservations = len(observations)
    A = parameters[0]
    C = 1
    Q = parameters[1]**2
    R = parameters[2]**2

    predictiveCov = initialStateCov
    xHatPredicted = initialState * np.ones((noObservations + 1, 1))
    xHatFiltered = initialState * np.ones((noObservations, 1))

    for t in range(0, noObservations):
        # Correction step
        S = C * predictiveCov * C + R
        kalmanGain = predictiveCov * C / S
        filteredCovariance = predictiveCov - kalmanGain * S * kalmanGain
        yHatPredicted = C * xHatPredicted[t]    
        xHatFiltered[t] = xHatPredicted[t] + kalmanGain * (observations[t - 1] - yHatPredicted)

        # Prediction step
        xHatPredicted[t + 1] = A * xHatFiltered[t]
        predictiveCov = A * filteredCovariance * A + Q

    return xHatFiltered

##############################################################################
# Fully-adapted particle filter for the linear Gaussian SSM
##############################################################################
def particleFilter(observations, parameters, noParticles, initialState):
        
    noObservations = len(observations) - 1
    phi = parameters[0]
    sigmav = parameters[1]
    sigmae = parameters[2]

    particles = np.zeros((noParticles, noObservations))
    ancestorIndices = np.zeros((noParticles, noObservations))
    weights = np.zeros((noParticles, noObservations))
    normalisedWeights = np.zeros((noParticles, noObservations))
    xHatFiltered = np.zeros((noObservations, 1))

    # Set the initial state and weights
    ancestorIndices[: , 0] = range(noParticles)
    particles[:, 0] = initialState
    xHatFiltered[0] = initialState
    normalisedWeights[:, 0] = 1.0 / noParticles
    logLikelihood = 0

    for t in range(1, noObservations):
        # Resample (multinomial)
        newAncestors = choice(noParticles, noParticles, p=normalisedWeights[:, t - 1], replace=True)
        ancestorIndices[:, 1:t - 1] = ancestorIndices[newAncestors, 1:t - 1]
        ancestorIndices[:, t] = newAncestors

        # Propagate
        part1 = (sigmav**(-2) + sigmae**(-2))**(-1)
        part2 = sigmae**(-2) * observations[t]
        part2 = part2 + sigmav**(-2) * phi * particles[newAncestors, t - 1]
        particles[:, t] = part1 * part2 + np.sqrt(part1) * randn(1, noParticles)

        # Compute weights
        yhatMean = phi * particles[:, t]
        yhatVariance = np.sqrt(sigmav**2 + sigmae**2)
        weights[:, t] = norm.logpdf(observations[t + 1], yhatMean, yhatVariance)

        maxWeight = np.max(weights[:, t])
        weights[:, t] = np.exp(weights[:, t] - maxWeight)
        sumWeights = np.sum(weights[:, t])
        normalisedWeights[:, t] = weights[:, t] / sumWeights

        # Estimate the state
        xHatFiltered[t] = np.sum(normalisedWeights[:, t] * particles[:, t])

        # Estimate log-likelihood
        predictiveLikelihood = maxWeight + np.log(sumWeights) - np.log(noParticles)
        logLikelihood += predictiveLikelihood

    return xHatFiltered, logLikelihood

##############################################################################
# Bootstrap particle filter for the stochastic volatility model
##############################################################################
def particleFilterSVmodel(observations, parameters, noParticles):

    noObservations = len(observations)
    mu = parameters[0]
    phi = parameters[1]
    sigmav = parameters[2]

    particles = np.zeros((noParticles, noObservations))
    ancestorIndices = np.zeros((noParticles, noObservations))
    weights = np.zeros((noParticles, noObservations))
    normalisedWeights = np.zeros((noParticles, noObservations))
    xHatFiltered = np.zeros((noObservations, 1))

    # Set the initial state and weights
    particles[:, 0] = mu + sigmav / np.sqrt(1.0 - phi**2) * randn(1, noParticles)
    normalisedWeights[:, 0] = 1.0 / noParticles
    weights[:, 0] = 1.0
    logLikelihood = 0
    
    for t in range(1, noObservations):
        # Resample particles
        newAncestors = choice(noParticles, noParticles, p=normalisedWeights[:, t - 1], replace=True)
        ancestorIndices[:, 1:t - 1] = ancestorIndices[newAncestors, 1:t - 1]
        ancestorIndices[:, t] = newAncestors

        # Propagate particles
        particles[:, t] = mu + phi * (particles[newAncestors, t - 1] - mu) + sigmav * randn(1, noParticles)

        # Weight particles
        weights[:, t] = norm.logpdf(observations[t - 1], 0, np.exp(particles[:, t] / 2))

        maxWeight = np.max(weights[:, t])
        weights[:, t] = np.exp(weights[:, t] - maxWeight)
        sumWeights = np.sum(weights[:, t])
        normalisedWeights[:, t] = weights[:, t] / sumWeights

        # Estimate the filtered state
        xHatFiltered[t] = np.sum(normalisedWeights[:, t] * particles[:, t])

        # Estimate log-likelihood
        predictiveLikelihood = maxWeight + np.log(sumWeights) - np.log(noParticles)
        logLikelihood += predictiveLikelihood

    
    # Sample the state estimate using the weights at t=T
    ancestorIndex = choice(noParticles, 1, p=normalisedWeights[:, noObservations - 1])
    stateTrajectory = particles[ancestorIndices[ancestorIndex, noObservations - 1].astype(int), :]

    return stateTrajectory, logLikelihood

File: r/helpers/stateEstimation.R
##############################################################################
# State estimation in LGSS and SV models using Kalman and particle filters.
#
# Johan Dahlin <liu (at) johandahlin.com.nospam>
# Documentation at https://github.com/compops/pmh-tutorial
# Published under GNU General Public License
##############################################################################

##############################################################################
# Fully-adapted particle filter for the linear Gaussian SSM
##############################################################################
particleFilter <- function(y, theta, noParticles, initialState) {
  
  T <- length(y) - 1
  phi <- theta[1] 
  sigmav <- theta[2]
  sigmae <- theta[3]
  
  # Initialise variables
  particles <- matrix(0, nrow = noParticles, ncol = T + 1)
  ancestorIndices <- matrix(0, nrow = noParticles, ncol = T + 1)
  weights <- matrix(1, nrow = noParticles, ncol = T + 1)
  normalisedWeights <- matrix(0, nrow = noParticles, ncol = T + 1)
  xHatFiltered <- matrix(0, nrow = T, ncol = 1)
  logLikelihood <- 0
  
  ancestorIndices[, 1] <- 1:noParticles
  particles[ ,1] <- initialState
  xHatFiltered[ ,1] <- initialState  
  normalisedWeights[, 1] = 1 / noParticles
  
  for (t in 2:T) {
    # Resample ( multinomial )
    newAncestors <- sample(noParticles, replace = TRUE, prob = normalisedWeights[, t - 1])
    ancestorIndices[, 1:(t - 1)] <- ancestorIndices[newAncestors, 1:(t - 1)]
    ancestorIndices[, t] <- newAncestors
    
    # Propagate
    part1 <- (sigmav^(-2) + sigmae^(-2))^(-1)
    part2 <- sigmae^(-2) * y[t]
    part2 <- part2 + sigmav^(-2) * phi * particles[newAncestors, t - 1]
    particles[, t] <- part1 * part2 + rnorm(noParticles, 0, sqrt(part1))
    
    # Compute weights
    yhatMean <- phi * particles[, t]
    yhatVariance <- sqrt(sigmae^2 + sigmav^2)
    weights[, t] <- dnorm(y[t + 1], yhatMean, yhatVariance, log = TRUE)
    
    maxWeight <- max(weights[, t])
    weights[, t] <- exp(weights[, t] - maxWeight)
    
    sumWeights <- sum(weights[, t])
    normalisedWeights[, t] <- weights[, t] / sumWeights
    
    # Estimate the state
    xHatFiltered[t] <- mean(particles[, t])
    
    # Estimate the log-likelihood
    predictiveLikelihood <- maxWeight + log(sumWeights) - log(noParticles)
    logLikelihood <- logLikelihood + predictiveLikelihood
    
  }

  list(xHatFiltered = xHatFiltered,
       logLikelihood = logLikelihood,
       particles = particles,
       weights = normalisedWeights)
  
}

##############################################################################
# Kalman filter for the linear Gaussian SSM
##############################################################################
kalmanFilter <- function(y, theta, initialState, initialStateCovariance) {
  
  T <- length(y)
  yHatPredicted <- matrix(initialState, nrow = T, ncol = 1)
  xHatFiltered <- matrix(initialState, nrow = T, ncol = 1)
  xHatPredicted <- matrix(initialState, nrow = T + 1, ncol = 1)
  predictedStateCovariance <- initialStateCovariance
  logLikelihood <- 0
  
  A <- theta[1] 
  C <- 1
  Q <- theta[2] ^ 2
  R <- theta[3] ^ 2
  
  for (t in 2:T) {
    # Correction step
    S <- C * predictedStateCovariance * C + R
    kalmanGain <- predictedStateCovariance * C / S
    filteredStateCovariance <- predictedStateCovariance - kalmanGain * S * kalmanGain
    
    yHatPredicted[t] <- C * xHatPredicted[t]
    xHatFiltered[t] <- xHatPredicted[t] + kalmanGain * (y[t] - yHatPredicted[t])
    
    # Prediction step
    xHatPredicted[t + 1] <- A * xHatFiltered[t]
    predictedStateCovariance <- A * filteredStateCovariance * A + Q
    
    # Estimate loglikelihood (not in the last iteration, to be able to compare with faPF)
    if (t < T) {
      logLikelihood = logLikelihood + dnorm(y[t], yHatPredicted[t], sqrt(S), log = TRUE)
    }
  }
  
  list(xHatFiltered = xHatFiltered, logLikelihood = logLikelihood)
}

##############################################################################
# Bootstrap particle filter for the stochastic volatility model
##############################################################################
particleFilterSVmodel <- function(y, theta, noParticles) {
  
  T <- length(y) - 1
  mu <- theta[1] 
  phi <- theta[2]
  sigmav <- theta[3]  
  
  particles <- matrix(0, nrow = noParticles, ncol = T + 1)
  ancestorIndices <- matrix(0, nrow = noParticles, ncol = T + 1)
  weights <- matrix(1, nrow = noParticles, ncol = T + 1)
  normalisedWeights <- matrix(0, nrow = noParticles, ncol = T + 1)
  xHatFiltered <- matrix(0, nrow = T, ncol = 1)
  logLikelihood <- 0
  
  ancestorIndices[, 1] <- 1:noParticles
  normalisedWeights[, 1] = 1 / noParticles

  # Generate initial state
  particles[, 1] <- rnorm(noParticles, mu, sigmav / sqrt(1 - phi^2))
  
  for (t in 2:(T + 1)) {
    # Resample ( multinomial )
    newAncestors <- sample(noParticles, replace = TRUE, prob = normalisedWeights[, t - 1])
    ancestorIndices[, 1:(t - 1)] <- ancestorIndices[newAncestors, 1:(t - 1)]
    ancestorIndices[, t] <- newAncestors
    
    # Propagate
    part1 <- mu + phi * (particles[newAncestors, t - 1] - mu)
    particles[, t] <- part1 + rnorm(noParticles, 0, sigmav)
    
    # Compute weights
    yhatMean <- 0
    yhatVariance <- exp(particles[, t] / 2)
    weights[, t] <- dnorm(y[t - 1], yhatMean, yhatVariance, log = TRUE)
    
    maxWeight <- max(weights[, t])
    weights[, t] <- exp(weights[, t] - maxWeight)
    
    sumWeights <- sum(weights[, t])
    normalisedWeights[, t] <- weights[, t] / sumWeights
    
    # Estimate the log-likelihood
    logLikelihood <- logLikelihood + maxWeight + log(sumWeights) - log(noParticles)
    
  }

  # Sample the state estimate using the weights at t=T
  ancestorIndex  <- sample(noParticles, 1, prob = normalisedWeights[, T])
  xHatFiltered <- particles[cbind(ancestorIndices[ancestorIndex, ], 1:(T + 1))]
  
  list(xHatFiltered = xHatFiltered, logLikelihood = logLikelihood)
}